feat: optimize gas usage

zsystm opened the pull request:
## Description

* Closes: #33 
* fix: set at least one validator when create randomized state. If not, there will be an error by ethermint side (no coinbase error when load evm config)
* feat: bump sdk version from v0.45.9-2 to v0.45.9-4 (diff: https://github.com/b-harvest/cosmos-sdk/compare/v0.45.9-2-canto-lsm-sim...v0.45.9-4-canto-lsm-sim)
* chore: change variable name (nas --> nase)

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] targeted the correct branch (see [PR Targeting](https://github.com/canto-network/canto/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] included the necessary unit and integration tests
- [x] reviewed "Files changed" and left comments if necessary
 
### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

zsystm committed:
tmp: add logger to analyze gas consumption

logs gas consumption at fileLogger

zsystm committed:
disable minus fee rate

zsystm committed:
chore: remove un-used errors and bump cosmos-sdk

zsystm committed:
fix duplicate error code

zsystm committed:
update simop

if calculated fee rate is negative, then apply zeroFeeRate

zsystm committed:
Merge remote-tracking branch 'origin/zsystm/fix-minus-fee-rate' into zsystm/lsm-gas-optimization

zsystm committed:
limit minting if it exceeds mainnet total supply

zsystm committed:
update: initialize at least one validator in AppStateRandomizedFn

zsystm committed:
apply Canto mainnet supply

based on mainnet supply, rich account get fund when initialize genesis on simulation testing.

each operation which needs token should get fund from rich account.

zsystm committed:
update CalcNetAmountStateEssentials: return multiple data structures

this is for optimizing gas consumption when QueueLiquidUnstake

instead of Iterate chunks separately, iterate it once when calc net amount state essentials.

zsystm committed:
refund gas used by cachedCtx

zsystm committed:
remove logger and print lines for debugging

zsystm committed:
Merge branch 'liquidstaking-module' into zsystm/lsm-gas-optimization

zsystm committed:
remove un-necessary lines

zsystm committed:
chore: change variable name

dongsam commented on x/liquidstaking/types/insurance.go:
```diff
@@ -55,8 +55,8 @@ func SortInsurances(validatorMap map[string]stakingtypes.Validator, insurances [
 
-		iValidator := validatorMap[iInsurance.ValidatorAddress]
-		jValidator := validatorMap[jInsurance.ValidatorAddress]
+		iValidator := validatorMap[iInsurance.GetValidator().String()]
+		jValidator := validatorMap[jInsurance.GetValidator().String()]
```
Is there a reason why you had to change it like this?

zsystm replied:
@dongsam 

No special reason for this. 
I just thought `GetValidator().String()` is more uniform than `insurance.ValidatorAddress` because when create validatorMap in `GetNetAmountStateEssentials`, it uses `GetValidator().String()` as a key. 

But, I saw that both `GetValidator().String()` and `insurance.ValidatorAddress` are used in liquidstaking module's code.
Both values are same, so it would be great to unify that usage in code base for readability.

dongsam replied:
[`GetValidator()`](https://github.com/Canto-Network/Canto/blob/b6c34839eb021a68a55ab7eb936a88465145ab3f/x/liquidstaking/types/insurance.go#L43-L46) eventually computes using `insurance.ValidatorAddress`, and already has the resulting string in the `insurance.ValidatorAddress`,  it doesn't seem necessary to compute it again

zsystm replied:
@dongsam 
Oh.. I see why u comment for it. 
I got it! Thanks for the detail.

I'll check other `GetValidator` usages too and fix that(=duplicated computation). Thanks!

zsystm replied:
@dongsam 
updated it! https://github.com/b-harvest/Canto/pull/39/commits/8b66c5c4ec114e8d6dd189dda6cc4fff39efa6b9 

the reason why tests failed yesterday is that I changed [the line](https://github.com/b-harvest/Canto/pull/39/commits/8b66c5c4ec114e8d6dd189dda6cc4fff39efa6b9#diff-ab2f09d261b7e9f72f62052a1d86a36ce8cff3b911bb79243a8d7c603c866d9bR1609) to `!=`. It must be equal, so the logic was broken. But now it is fixed (using `==`).

zsystm committed:
reduce computation

GetValidator().String is a redundant computation. We can just use ValidatorAddress field.
